File: openkoto-ios/Packages/OpenKotoKit/Sources/OKSRS/FSRS.swift
import Foundation
import OKModels

/// FSRS-6 调度引擎(长期模式,天粒度)。
///
/// 1:1 移植自桌面 `textlingo-desktop/src-tauri/src/fsrs.rs`;
/// 跨端契约见 `docs/specs/vocabulary-srs-spec.md` §2。
/// 与 Rust 引擎必须通过同一份黄金用例 `Tests/OKSRSTests/Fixtures/fsrs_golden_v1.json`
/// (权威文件在 `docs/specs/fixtures/`,Rust 侧测试保证两份逐字节一致)。
/// 求值顺序与 round8 出现位置对齐参考实现 ts-fsrs 5.4.1,不得调整。
public enum FSRS {
    /// ts-fsrs 5.4.1 default_w(w0..w20)
    public static let defaultParams: [Double] = [
        0.212, 1.2931, 2.3065, 8.2956, 6.4133, 0.8334, 3.0194, 0.001,
        1.8722, 0.1666, 0.796, 1.4835, 0.0614, 0.2629, 1.6483, 0.6014,
        1.8729, 0.5425, 0.0912, 0.0658, 0.1542,
    ]

    public static let schedulerVersion = "fsrs6"
    public static let defaultDesiredRetention = 0.9
    private static let sMin = 0.001
    private static let maxInterval = 36500
    private static let sm2Retention = 0.9

    /// 评分档位:UI 三档映射 不认识→again,模糊→hard,认识→good;easy 引擎支持、当前 UI 不用。
    public enum Grade: Int, Sendable, CaseIterable {
        case again = 1
        case hard = 2
        case good = 3
        case easy = 4
    }

    public struct Update: Sendable, Equatable {
        public var stability: Double
        public var difficulty: Double
        public var intervalDays: Int
        public var state: SRSState
    }

    public enum EngineError: Error, Equatable {
        case invalidElapsedDays(Int)
        case invalidDesiredRetention(Double)
    }

    // MARK: - 数值约定(规范 §2.2)

    private static func round8(_ x: Double) -> Double {
        (x * 1e8).rounded() / 1e8
    }

    private static func clamp(_ x: Double, _ lo: Double, _ hi: Double) -> Double {
        min(max(x, lo), hi)
    }

    private static var decay: Double { -defaultParams[20] }

    private static var factor: Double { round8(pow(0.9, 1.0 / decay) - 1.0) }

    // MARK: - 保持率与间隔(规范 §2.3)

    /// R(t, S)。t 为距上次复习的天数(允许小数);t=0 → 1。
    public static func retrievability(stability: Double, elapsedDays: Double) -> Double {
        guard stability > 0 else { return 0 }
        return round8(pow(1.0 + factor * elapsedDays / stability, decay))
    }

    private static func intervalModifier(_ desiredRetention: Double) throws -> Double {
        guard desiredRetention > 0, desiredRetention <= 1 else {
            throw EngineError.invalidDesiredRetention(desiredRetention)
        }
        return round8((pow(desiredRetention, 1.0 / decay) - 1.0) / factor)
    }

    private static func nextInterval(stability: Double, modifier: Double) -> Int {
        let raw = max(1.0, (stability * modifier).rounded())
        return min(Int(raw), maxInterval)
    }

    // MARK: - 记忆状态更新(规范 §2.4)

    private static func initStability(_ grade: Grade) -> Double {
        max(defaultParams[grade.rawValue - 1], 0.1)
    }

    /// 注意:仅在 new 卡初始化路径 clamp 到 [1,10];meanReversion 里用的
    /// initDifficulty(.easy) 保持原值(可为负),与参考实现一致。
    private static func initDifficulty(_ grade: Grade) -> Double {
        let w = defaultParams
        return round8(w[4] - exp(Double(grade.rawValue - 1) * w[5]) + 1.0)
    }

    private static func linearDamping(deltaD: Double, oldD: Double) -> Double {
        round8(deltaD * (10.0 - oldD) / 9.0)
    }

    private static func meanReversion(initial: Double, current: Double) -> Double {
        let w = defaultParams
        return round8(w[7] * initial + (1.0 - w[7]) * current)
    }

    private static func nextDifficulty(_ difficulty: Double, _ grade: Grade) -> Double {
        let w = defaultParams
        let deltaD = -w[6] * (Double(grade.rawValue) - 3.0)
        let nextD = difficulty + linearDamping(deltaD: deltaD, oldD: difficulty)
        return clamp(meanReversion(initial: initDifficulty(.easy), current: nextD), 1.0, 10.0)
    }

    private static func nextRecallStability(
        _ difficulty: Double, _ stability: Double, _ r: Double, _ grade: Grade
    ) -> Double {
        let w = defaultParams
        let hardPenalty = grade == .hard ? w[15] : 1.0
        let easyBonus = grade == .easy ? w[16] : 1.0
        let grown =
            stability
            * (1.0
                + exp(w[8])
                    * (11.0 - difficulty)
                    * pow(stability, -w[9])
                    * (exp((1.0 - r) * w[10]) - 1.0)
                    * hardPenalty
                    * easyBonus)
        return round8(clamp(grown, sMin, Double(maxInterval)))
    }

    private static func nextForgetStability(
        _ difficulty: Double, _ stability: Double, _ r: Double
    ) -> Double {
        let w = defaultParams
        let forgotten =
            w[11]
            * pow(difficulty, -w[12])
            * (pow(stability + 1.0, w[13]) - 1.0)
            * exp((1.0 - r) * w[14])
        return round8(clamp(forgotten, sMin, Double(maxInterval)))
    }

    /// 单档位记忆状态更新(规范 §2.4)。
    private static func nextMemoryState(
        stability: Double, difficulty: Double, elapsedDays: Int, grade: Grade
    ) -> (stability: Double, difficulty: Double) {
        if stability == 0, difficulty == 0 {
            return (initStability(grade), clamp(initDifficulty(grade), 1.0, 10.0))
        }
        let r = retrievability(stability: stability, elapsedDays: Double(elapsedDays))
        let newS: Double
        if grade == .again {
            // 长期模式失败:S ← min(S, S_forget)
            newS = clamp(round8(stability), sMin, nextForgetStability(difficulty, stability, r))
        } else {
            newS = nextRecallStability(difficulty, stability, r, grade)
        }
        return (newS, nextDifficulty(difficulty, grade))
    }

    // MARK: - 一次复习(规范 §2.5)

    /// 同时计算四档并做跨档位间隔排序修正后取所选档。
    public static func nextReview(
        stability: Double,
        difficulty: Double,
        elapsedDays: Int,
        grade: Grade,
        desiredRetention: Double = FSRS.defaultDesiredRetention
    ) throws -> Update {
        guard elapsedDays >= 0 else { throw EngineError.invalidElapsedDays(elapsedDays) }
        let modifier = try intervalModifier(desiredRetention)

        var states: [(stability: Double, difficulty: Double)] = []
        var intervals: [Int] = []
        for g in Grade.allCases {
            let state = nextMemoryState(
                stability: stability, difficulty: difficulty, elapsedDays: elapsedDays, grade: g)
            states.append(state)
            intervals.append(nextInterval(stability: state.stability, modifier: modifier))
        }
        // 排序修正(顺序执行,与参考实现一致)
        intervals[0] = min(intervals[0], intervals[1])
        intervals[1] = max(intervals[1], intervals[0] + 1)
        intervals[2] = max(intervals[2], intervals[1] + 1)
        intervals[3] = max(intervals[3], intervals[2] + 1)

        let index = grade.rawValue - 1
        return Update(
            stability: states[index].stability,
            difficulty: states[index].difficulty,
            intervalDays: intervals[index],
            state: grade == .again ? .learning : .review
        )
    }

    // MARK: - SM-2 种子(规范 §4;iOS 无存量用户,仅为跨端一致性保留)

    public static func seedFromSM2(
        intervalDays: Int, easeFactor: Double
    ) -> (stability: Double, difficulty: Double) {
        let w = defaultParams
        let stability = max(Double(intervalDays), 0.1) / (9.0 * (1.0 / sm2Retention - 1.0))
        let denominator =
            exp(w[8]) * pow(stability, -w[9]) * (exp((1.0 - sm2Retention) * w[10]) - 1.0)
        let difficulty = clamp(11.0 - (easeFactor - 1.0) / denominator, 1.0, 10.0)
        return (stability, difficulty)
    }
}

File: textlingo-desktop/src-tauri/src/fsrs.rs
//! FSRS-6 调度引擎(长期模式,天粒度)。
//!
//! 跨端契约:docs/specs/vocabulary-srs-spec.md §2。
//! 与 openkoto-ios 的 OKSRS/FSRS.swift 必须通过同一份黄金用例
//! docs/specs/fixtures/fsrs_golden_v1.json(见 tests/fsrs_golden_test.rs)。
//! 求值顺序与 round8 出现位置对齐参考实现 ts-fsrs 5.4.1,不得调整。

pub const FSRS6_DEFAULT_PARAMS: [f64; 21] = [
    0.212, 1.2931, 2.3065, 8.2956, 6.4133, 0.8334, 3.0194, 0.001, 1.8722, 0.1666, 0.796, 1.4835,
    0.0614, 0.2629, 1.6483, 0.6014, 1.8729, 0.5425, 0.0912, 0.0658, 0.1542,
];

pub const SCHEDULER_VERSION: &str = "fsrs6";
pub const DEFAULT_DESIRED_RETENTION: f64 = 0.9;
const S_MIN: f64 = 0.001;
const MAX_INTERVAL: i32 = 36500;
const SM2_RETENTION: f64 = 0.9;

#[derive(Debug, Clone, PartialEq)]
pub struct FsrsUpdate {
    pub stability: f64,
    pub difficulty: f64,
    pub interval_days: i32,
    pub srs_state: String,
}

fn round8(x: f64) -> f64 {
    (x * 1e8).round() / 1e8
}

fn clamp(x: f64, lo: f64, hi: f64) -> f64 {
    x.max(lo).min(hi)
}

fn decay() -> f64 {
    -FSRS6_DEFAULT_PARAMS[20]
}

fn factor() -> f64 {
    round8(0.9_f64.powf(1.0 / decay()) - 1.0)
}

/// 保持率 R(t, S)。t 为距上次复习的天数(允许小数);t=0 → 1。
pub fn retrievability(stability: f64, elapsed_days: f64) -> f64 {
    if stability <= 0.0 {
        return 0.0;
    }
    round8((1.0 + factor() * elapsed_days / stability).powf(decay()))
}

fn interval_modifier(desired_retention: f64) -> Result<f64, String> {
    if desired_retention <= 0.0 || desired_retention > 1.0 {
        return Err(format!(
            "Invalid desired_retention {desired_retention}, expected (0, 1]"
        ));
    }
    Ok(round8(
        (desired_retention.powf(1.0 / decay()) - 1.0) / factor(),
    ))
}

fn next_interval(stability: f64, modifier: f64) -> i32 {
    let raw = (stability * modifier).round().max(1.0);
    (raw as i32).min(MAX_INTERVAL)
}

fn init_stability(grade: u8) -> f64 {
    FSRS6_DEFAULT_PARAMS[(grade - 1) as usize].max(0.1)
}

/// 注意:仅在 new 卡初始化路径 clamp 到 [1,10];mean_reversion 里用的
/// init_difficulty(4) 保持原值(可为负),与参考实现一致。
fn init_difficulty(grade: u8) -> f64 {
    let w = &FSRS6_DEFAULT_PARAMS;
    round8(w[4] - (((grade - 1) as f64) * w[5]).exp() + 1.0)
}

fn linear_damping(delta_d: f64, old_d: f64) -> f64 {
    round8(delta_d * (10.0 - old_d) / 9.0)
}

fn mean_reversion(init: f64, current: f64) -> f64 {
    let w = &FSRS6_DEFAULT_PARAMS;
    round8(w[7] * init + (1.0 - w[7]) * current)
}

fn next_difficulty(difficulty: f64, grade: u8) -> f64 {
    let w = &FSRS6_DEFAULT_PARAMS;
    let delta_d = -w[6] * ((grade as f64) - 3.0);
    let next_d = difficulty + linear_damping(delta_d, difficulty);
    clamp(mean_reversion(init_difficulty(4), next_d), 1.0, 10.0)
}

fn next_recall_stability(difficulty: f64, stability: f64, r: f64, grade: u8) -> f64 {
    let w = &FSRS6_DEFAULT_PARAMS;
    let hard_penalty = if grade == 2 { w[15] } else { 1.0 };
    let easy_bonus = if grade == 4 { w[16] } else { 1.0 };
    round8(clamp(
        stability
            * (1.0
                + w[8].exp()
                    * (11.0 - difficulty)
                    * stability.powf(-w[9])
                    * (((1.0 - r) * w[10]).exp() - 1.0)
                    * hard_penalty
                    * easy_bonus),
        S_MIN,
        MAX_INTERVAL as f64,
    ))
}

fn next_forget_stability(difficulty: f64, stability: f64, r: f64) -> f64 {
    let w = &FSRS6_DEFAULT_PARAMS;
    round8(clamp(
        w[11]
            * difficulty.powf(-w[12])
            * ((stability + 1.0).powf(w[13]) - 1.0)
            * (((1.0 - r) * w[14]).exp()),
        S_MIN,
        MAX_INTERVAL as f64,
    ))
}

/// 单档位记忆状态更新(规范 §2.4)。
fn next_memory_state(stability: f64, difficulty: f64, elapsed_days: i64, grade: u8) -> (f64, f64) {
    if stability == 0.0 && difficulty == 0.0 {
        return (
            init_stability(grade),
            clamp(init_difficulty(grade), 1.0, 10.0),
        );
    }
    let r = retrievability(stability, elapsed_days as f64);
    let new_s = if grade == 1 {
        // 长期模式失败:S ← min(S, S_forget)
        clamp(
            round8(stability),
            S_MIN,
            next_forget_stability(difficulty, stability, r),
        )
    } else {
        next_recall_stability(difficulty, stability, r, grade)
    };
    (new_s, next_difficulty(difficulty, grade))
}

/// 一次复习(规范 §2.5):同时计算四档并做跨档位间隔排序修正后取所选档。
pub fn next_review(
    stability: f64,
    difficulty: f64,
    elapsed_days: i64,
    grade: u8,
    desired_retention: f64,
) -> Result<FsrsUpdate, String> {
    if !(1..=4).contains(&grade) {
        return Err(format!("Invalid grade {grade}, expected 1..=4"));
    }
    if elapsed_days < 0 {
        return Err(format!("Invalid elapsed_days {elapsed_days}"));
    }
    let modifier = interval_modifier(desired_retention)?;

    let mut states = [(0.0_f64, 0.0_f64); 4];
    let mut intervals = [0_i32; 4];
    for g in 1..=4u8 {
        let state = next_memory_state(stability, difficulty, elapsed_days, g);
        intervals[(g - 1) as usize] = next_interval(state.0, modifier);
        states[(g - 1) as usize] = state;
    }
    // 排序修正(顺序执行,与参考实现一致)
    intervals[0] = intervals[0].min(intervals[1]);
    intervals[1] = intervals[1].max(intervals[0] + 1);
    intervals[2] = intervals[2].max(intervals[1] + 1);
    intervals[3] = intervals[3].max(intervals[2] + 1);

    let idx = (grade - 1) as usize;
    Ok(FsrsUpdate {
        stability: states[idx].0,
        difficulty: states[idx].1,
        interval_days: intervals[idx],
        srs_state: if grade == 1 { "learning" } else { "review" }.to_string(),
    })
}

/// SM-2 → FSRS 一次性种子(规范 §4,来源 fsrs-rs memory_state_from_sm2)。
pub fn seed_from_sm2(interval_days: i32, ease_factor: f64) -> (f64, f64) {
    let w = &FSRS6_DEFAULT_PARAMS;
    let stability = (interval_days as f64).max(0.1) / (9.0 * (1.0 / SM2_RETENTION - 1.0));
    let denominator =
        w[8].exp() * stability.powf(-w[9]) * (((1.0 - SM2_RETENTION) * w[10]).exp() - 1.0);
    let difficulty = clamp(11.0 - (ease_factor - 1.0) / denominator, 1.0, 10.0);
    (stability, difficulty)
}

/// UI 三档评分 → 引擎档位:不认识→1(Again),模糊→2(Hard),认识→3(Good)。
pub fn grade_from_str(grade: &str) -> Result<u8, String> {
    match grade {
        "unknown" => Ok(1),
        "uncertain" => Ok(2),
        "known" => Ok(3),
        _ => Err("Invalid grade, expected unknown|uncertain|known".to_string()),
    }
}
